### ISSUE TITLE
ENH: Future of npymath library

### ISSUE THREAD

**Issue by matthew-brett** [numpy/numpy#20880]
### Proposed new feature or change:

This is an issue to discuss the future of the `npymath` library in Numpy.

Numpy ships with a static library in `core/lib` named `npymath.lib` (Windows) or `libnpymath.a` (Unices).

It provides a suitable platform-specific implementation of various math routines.

Scipy links to `npymath` during its build process, in various places.

We (Scipy) have run into problems linking against Npymath, because of the combination of Numpy's use of the latest Visual Studio toolchain, and Scipy's use of the Mingw-w64 toolchain - discussed in https://github.com/MacPython/numpy-wheels/issues/145 .

The issue that arises is that it is in general more difficult to link static libraries across toolchains, than it is to link against dynamic libraries (e.g. DLLs) or to recompile the sources.   This issue is to discuss whether we should think of other solutions.   Possible options are:

* Leave things as they are.
* Ship the C header and sources for Npymath, so external libraries and packages can recompile them.
* Create a dynamic library from Npymath, and ship that, for linkage against external libraries.
* Your option here.

This issue is to house the discussion of different options.

**Comment by rgommers** [comment]
Note that whether or not this is a priority is kinda coupled with the `vc141` vs. newer MS compiler toolchain discussion in https://github.com/MacPython/numpy-wheels/issues/145. For `vc141` it seems we can work with `libnpymath.a` in a Mingw-w64 based SciPy build, which with `vc142` things just seem broken and likely hard/fragile to repair.

We could do something proactively, but the urgency is a lot lower as long as there's no need to upgrade from `vc141`.

**Comment by carlkl** [comment]
my suggestion is as follows:

- leave `npymath.lib` and `npyrandom.lib` as they are for numpy.
- Numpy could create two additional shared binaries  `npymath.dll` and `npyrandom.dll` from these static libraries.
- during the Numpy installation, place these two DLLs in a common location where Scipy or other third-party packages can locate them both during the build process and during import.


This approach has the following benefits:

- no need to change the build process of Numpy as it is only an extension of the existing build process.
- packages which depends on these libraries can still use the static libraries if **and only if** they use the same VS toolset as the Numpy uses. 
- packages which depends on these libraries using a different toolset or a different compiler can link against the shared libraries instead.

For a test I have successfully build `npymath.dll` with VS `link.exe`. It only needs a def file with all symbol names to be exported. A test to build scipy with this change will follow. 

There is a disadvantage as well:

One has to consider the situation that someone installs an Numpy version without theses additional DLLs and then Scipy which needs them. Mitigation could be:

- build Scipy for windows with a different minimal Numpy version to avoid this situation.
- copy these DLLs to the Scipy wheel as well and load them on import.

**Comment by rgommers** [comment]
> Ship the C header and sources for Npymath, so external libraries and packages can recompile them.

Not a terrible idea in principle, but it would probably be as much work as shipping a DLL and permanently increase build times for libraries using that C code.

> Create a dynamic library from Npymath, and ship that, for linkage against external libraries.

This is probably best. I agree with @carlkl's suggestion's mostly, with two tweaks:
1. It should be done on all platforms and not just on Windows
2. The install location isn't arbitrary, we should install to libdir (`sysconfig.get_config_var('LIBDIR')`)

> build Scipy for windows with a different minimal Numpy version to avoid this situation.

This would be the better option - if you need a newer NumPy, then bump the minimum version.

**Comment by mattip** [comment]
I came across this again when building NumPy from source when a wheel was not available. It would be nice to solve this for 1.23 since moving forward more people will be hitting this when they compile NumPy and Scipy from source with a modern visual studio.

One complication is that on windows,  `sysconfig.get_config_var('LIBDIR')` does not exist. There is a DLL directory under `sysconfig.get_config_var('base') (which is the same as `sysconfig.get_path('base')`) which would be appropriate. The import lib `npymath.lib` needed for linking currently lives in `site-packages\numpy\core\lib`, so it can probably stay there.

**Comment by mattip** [comment]
I am becoming convinced that the DLLs directory in windows is not a good target. It is not added to the dll search path (see [os.add_dll_directory](https://docs.python.org/3/library/os.html#os.add_dll_directory)), rather it is added to `sys.path`. So any project linking to npymath would need to call `os.add_dll_directory(<path-to-DLLs>), and there is no canonical way to discover what that directory is. In fact on vanilla python from python.org, no directory is added to the dll search path, so there is no reason to prefer the DLLs directory over any random directory in site-packages. Are there any best-practices around this subject?

**Comment by rgommers** [comment]
> I came across this again when building NumPy from source when a wheel was not available.

How exactly? This issue was opened when we discovered it with the mingw-w64 compilers. What's your exact config here, and how does this turn up when building numpy only rather than linking `libnpymath` in another project?

> One complication is that on windows, `sysconfig.get_config_var('LIBDIR')` does not exist.

I'm not sure what `sysconfig.get_config_var('LIBDIR')` is supposed to do, so ignore that part of my suggestion. What I meant is: _the lib directory where the `libpython` for the Python version you are using is located_. You have to link that too, so it should be discoverable somehow.

**Comment by mattip** [comment]
> How exactly? 

See [this to build Scipy on PyPy](https://github.com/mattip/scipy/pull/1) using meson: it builds NumPy from source on the CI machine.

> You have to link that too

Linking is the easy part. The harder problem is to find the DLL at runtime. Like LD_LIBRARY_PATH on Linux, windows has an internal list of directories it will search. For the NumPy/SciPy OpenBLAS dll, we call `os.add_dll_directory` in `__config__.py`. If we ship `npmath.dll`, consumers would need to also add something like this to their code. https://github.com/numpy/numpy/blob/46ec71f6bc74a0a248036b305fade5562e396c52/numpy/distutils/misc_util.py#L2354-L2359.

**Comment by matthew-brett** [comment]
I would like to return to the idea of shipping the C sources.  Perhaps that would just be as simple as copying the generated `.c` and `.h` sources in the `npymath` directory into a directory, such as `<np-include>/npymath`.

I think that would be easy to maintain - in practice, surely easier than working out a general mechanism for Numpy to export a DLL directory?   Or do we need that for some other reason?

**Comment by carlkl** [comment]
I would propose to copy npmath.dll from numpy into scipy's `.libs` folder during the scipy build process. And to import npmath.dll from there during scipy import.

**Comment by carlkl** [comment]
Another proposal:

As npymath.dll is quite small (I tried this out it in januarý and got a 56kB sized npymath.dll) it could be a good idea to simply load this DLL during numpy import. Once the DLL is loaded into process space it will be available for all packages depending on this DLL. Symbols of this DLL does not interfere with the static linked symbols used by numpy, that means numpy will just ignore this DLL.

**Comment by rgommers** [comment]
> I think that would be easy to maintain - in practice, surely easier than working out a general mechanism for Numpy to export a DLL directory? Or do we need that for some other reason?

Looking at the code to build `npymath` in https://github.com/numpy/numpy/blob/main/numpy/core/setup.py, it's not exactly trivial with compiler flags, depending on math libs, etc. (although consumers may already have similar code in their own build setup, not sure). Also, note that we have to give `npyrandom` the same treatment.

I think I like the DLL plan a little better.

> As npymath.dll is quite small (I tried this out it in januarý and got a 56kB sized npymath.dll) it could be a good idea to simply load this DLL during numpy import.

Of the three ways of using a DLL, this is probably the nicest one - assuming we won't run into trouble with changes in the library & versioning. `libnpyrandom.a` is ~25% larger than `libnpymath.a`, so that's still okay.

**Comment by matthew-brett** [comment]
I agree - that's an excellent idea - preloading the DLL.   Would this be the right approach for npyrandom?  Or is that more likely to be compiled into projects that don't import Numpy?

**Comment by rgommers** [comment]
I can't find many users, but SciPy uses both, and https://github.com/Erotemic/xdev/commit/77a571298529d1e89e0285e6aaa763a6bc849849#diff-ffb8093dfe715993aed3918bfa9b1fb2e59c13d4a5699aab5088df611f960209R230 is an example of use in a random project I hadn't seen before (still uses `numpy`). So I'd say use cases are similar.

**Comment by mattip** [comment]
The problem with injecting npymath.dll via `import numpy` is that, well, it requires `import numpy`. I hope all projects using npymath import numpy. Building as a shared library seems like a reasonable first step.

I do wonder how far we can go to make npymath more of a header-file-only library, along the lines of eigen. The code in [npy_math_internal.h.src](https://github.com/numpy/numpy/blob/main/numpy/core/src/npymath/npy_math_internal.h.src)](https://github.com/numpy/numpy/tree/main/numpy/core/src/npymath), could be shipped as a header file if we always use the `NPY_INLINE_MATH` macro. Complex number support has [improved in MSVC](https://docs.microsoft.com/en-us/cpp/c-runtime-library/complex-math-support?view=msvc-170) so we may be able to alias native routines instead of the hand written ones in [npy_math_complex.c.src](https://github.com/numpy/numpy/blob/main/numpy/core/src/npymath/npy_math_complex.c.src). And it seems like there is work that can be done to define more `HAVE_` macros like `HAVE_ASINH`. This can be a follow-on project, after we change the build to create and ship the dll.

**Comment by carlkl** [comment]
> Complex number support has improved in MSVC so we may be able to alias native routines instead of the hand written ones in npy_math_complex.c.src. 

Most trigonometric symbols (also the complex ones) are backed by `api-ms-win-crt-math-l1-1-0.dll`. See the attached: [npymath_api-ms-win-crt-math_imports.txt](https://github.com/numpy/numpy/files/8523990/npymath_api-ms-win-crt-math_imports.txt)

**Comment by seberg** [comment]
Also ping @serge-sans-paille since I am worried this affects gh-21487.

One thing I am unclear about is whether using a dynamically linked library here might be problematic for some users because it prevents inlining?  A lot of the functions are headers though, so maybe Matti's thought about just making it header-only makes sense?

Since `npymath` is effectively statically linked (or headers only).  I now do wonder whether it might make sense to split it out and include it as a submodule in NumPy?  That would also allow downstream to recieve bug-fixes by compiling against an old version of NumPy, but a new version of `npymath`?

**Comment by carlkl** [comment]
as described in https://github.com/numpy/numpy/issues/20880#issuecomment-1021097141, supplying static as well dynamic libraries for npymath and npyrandom could be a solution.

**Comment by h-vetinari** [comment]
To make the link bidirectional: conda-forge is currently discussing moving to vc142 (leaning towards implementation soon), perhaps some people here are interested or have an opinion.

**Comment by carlkl** [comment]
As it is trivial to build DLLs from the npymath.lib and the npyrandom.lib libraries I propose to use DLLs rather than the VC lib files for the scipy building process on windows with mingw-w64. It is even possible to build one single DLL for the sake of simplicity, which include the object code from both libraries.

This procedure was necessary for the windows 32-bit scipy-meson test build anyway.

The advantage of this procedure: it works regardless of the VC version. 
Disadvantage: this DLLs need to be integrated into the scipy wheel.

**Comment by rgommers** [comment]
I'm finding myself toying with the idea of just shipping the sources again. It's raining issues with `libnpymath.a`, with cross-compiling SciPy 1.9.0, and also with `cibuildwheel` builds for SciPy on Windows (the 3 releases we shipped with `vc142`).

> Looking at the code to build `npymath` in https://github.com/numpy/numpy/blob/main/numpy/core/setup.py, it's not exactly trivial with compiler flags, depending on math libs, etc. (although consumers may already have similar code in their own build setup, not sure)

In the end it's not that bad; a lot of that is weird handling of the math library. For SciPy we can probably just ignore all that. It would also help with solving the problem now - we look for the install source files, and if they're missing then we use a vendored copy. With a DLL, that doesn't work.

**Comment by carlkl** [comment]
Unfortunately one also needs then source to numpy random as well, as numpy.random also depends on npymath symbols.
All in all I like the DLL idea best, because it is the easiest way to escape the compiler mix hell.

Another idea would be to build numpy with mingw-w64 as well and use that to build scipy. A numpy build of this kind should follow the specifications of CPython (long double == double) and should use the windows umbrella libraries for the math symbols. Later on a scipy wheel build this way should work together with a standard PYPI MSVC build numpy without any problems IMHO. In the end it should be sufficient to exclusively build the minimum numpy version for a given scipy version.

**Comment by h-vetinari** [comment]
> Later on a scipy wheel build this way should work together with a standard PYPI MSVC build numpy without any problems IMHO.

If I understand the statement from the compiler devs in https://github.com/matthew-brett/dll_investigation/issues/1 correctly, this is not guaranteed to work, and we've been lucky enough that the subset of the C ABI we've been using was compatible between mingw and MSVC:

> If your code doesn't use anything else than plain C functions without complicated linkage, and doesn't call many tricky CRT functions, there's a chance it might work.
> 
> So overall, if mixing the two environments happens to work, good for you, but for developers of toolchains, it's pretty much a non-goal as it's not something that can be made to work in the general sense anyway.

**Comment by rgommers** [comment]
> If I understand the statement from the compiler devs in [matthew-brett/dll_investigation#1](https://github.com/matthew-brett/dll_investigation/issues/1) correctly, this is not guaranteed to work, and we've been lucky enough that the subset of the C ABI we've been using was compatible between mingw and MSVC:

Please let's not discuss that in this issue. We are relying on this for the Windows SciPy builds, have spent a lot of work on it already, and it is not directly related to the topic at hand here (`libnpymath`).

I'm giving vendoring a try right now, let's see how deep the rabbit hole is.

**Comment by carlkl** [comment]
We should avoid to build binaries (that is DLL's or pyd's) with code objects coming from different compiler toolchains. If this works, it works only by accident. But it was never a problem mixing DLLs in the same windows process space coming from different compilers as long as the DLL's or pyd's depends on the same CRT.

**Comment by rgommers** [comment]
> I'm giving vendoring a try right now, let's see how deep the rabbit hole is.

The good news: it took me 40 minutes to get everything to build from source within SciPy, and the test suite passes without relying on `libnpymath.a`.

The bad news: I cheated, and vendored also the generated `config.h` and `npy_cpu_dispatch_config.h` as well as a ton of other headers from `numpy/core/src/common`. Reimplementing all the feature checks is a lot of work, and not really feasible for most consumers (for SciPy it would be; the work to implement those checks in Meson has to be done anyway for NumPy itself). There's a lot that can be cleaned up in that process, but that takes a full release cycle and careful testing.

**Comment by mattip** [comment]
> Reimplementing all the feature checks is a lot of work

I think some of those feature checks and much of the functionality of npymath needs refactoring. Since we now require glibc2_17 (manylinux2014) and higher, and have dropped support for msvc<2017, we should check which of the [optional stdlib functions](https://github.com/numpy/numpy/blob/5028e407bf7c98662bfe247895292087e1372737/numpy/core/setup_common.py#L129) can be moved to mandatory and removed from npymath. The C99 variants of the mandatory functions I think also can be removed. Perhaps we could reduce the library to a header-only set of static functions.

**Comment by rgommers** [comment]
Agreed, it's in need of both cleaning up old stuff, C99 stuff and a good refactoring. The most hairy issue is not the number of system-provided checks, but the header spaghetti that drags in every header in `src/npymath` and `src/common`. I tried removing `npy_cpu_features.h` from that set for building `npymath`, but that was nontrivial.

**Comment by mattip** [comment]
> I tried removing npy_cpu_features.h from that set

numpy/numpy#22090 removes the npy_cpu* includes from `npy_config.h` (among other things)

**Comment by rgommers** [comment]
I am also trying to get a sense of what we actually still need from npymath in SciPy. In the end npymath was/is a C99 compatibility layer after all, and we can now rely on C99 (and C++11).

The only submodule we use it there is `scipy.special, and a lot of usage is only for nan/inf handling which we should now be able to get rid of. There's a small optimization in npymath where `npy_isnan` et al use `__builtin_isnan` instead of `isnan` if it's available (i.e., for GCC), but I'm not sure how relevant that is to performance in the big picture.

There's currently 444 lines of code where `npy_*` is used, so perhaps it's doable. https://github.com/scipy/scipy/pull/16800 makes a start.

**Comment by mattip** [comment]
+1 for eliminating it entirely, although I think it is still needed for the complex glibc routines that are buggy in glibc2.17 (manylinux2014).

**Comment by carlkl** [comment]
Unfortunately scipy depends on npyrandom, which itself depends on npymath as well.

**Comment by rgommers** [comment]
I'm not too worried about `npyrandom`, that's a dependency in exactly a single place (`scipy/stats/_biasedurn.pyx.templ`) which even still has the support code to build without `npyrandom`. So we can find a solution there.

> although I think it is still needed for the complex glibc routines that are buggy in glibc2.17

@mattip I can't find the details so quickly, and I don't see a mention of glibc in `core/src/npymath/npy_math_complex.c.src`. Do you remember what the problems were or where the relevant issue(s) live?

**Comment by mattip** [comment]
One way to get a feel for the issues is to look at the [blame for npy_config.h](https://github.com/numpy/numpy/blame/43c113dd7aa36ef833315035858ea98a7b4732c5/numpy/core/src/common/npy_config.h) which shows many of the PRs to blocklist functions

**Comment by mattip** [comment]
If you can run a manylinux2014 docker image, you can try unblocking the functions and see if tests pass. I don't think anything significant should have changed since #7518 though.

**Comment by rgommers** [comment]
Thanks, that helps. The glibc issues are pretty niche, and it looks none of those are used in `scipy.special`. The functions we're not using from MSVC (e.g. `clog`, `csqrt`) look more problematic.

**Comment by rgommers** [comment]
After some more digging, the issue is not just the complex routines, but that MSVC does not (and probably never will) support the C99 syntax (`complex float`, `complex double`), so `npy_cdouble` & co or an equivalent will remain necessary. That looks like the only thing though - if we could reduce `npymath` to those typedefs and the complex routines, that should be fairly concise and vendor-able.

**Comment by h-vetinari** [comment]
> That looks like the only thing though - if we could reduce `npymath` to those typedefs and the complex routines, that should be fairly concise and vendor-able.

I presume you've seen [this page](https://docs.microsoft.com/en-us/cpp/c-runtime-library/complex-math-support?view=msvc-170)? I start off with:
>The Microsoft C Runtime library (CRT) provides complex math library functions, including all of those required by ISO C99. The compiler doesn't directly support a `complex` or `_Complex` keyword, therefore the Microsoft implementation uses structure types to represent complex numbers.

So it should be possible to define `npy_cdouble` as a `complex` on unix and fix windows lack of C99 support behind the scenes?

**Comment by mattip** [comment]
> if we could reduce npymath to those typedefs and the complex routines, that should be fairly concise and vendor-able

> it should be possible to define npy_cdouble as a complex on unix and fix windows lack of C99 support behind the scenes

(Meme of multiple spidermen pointing to eachother) I think these are the same thing

**Comment by rgommers** [comment]
Yes indeed, I think `npy_cdouble` in its current form does the right thing already AFAICT. C11 and onwards decided all of `complex.h` is optional, so we probably should not make it Windows-specific, but instead don't assume it must exist anywhere.

**Comment by h-vetinari** [comment]
> C11 and onwards decided all of `complex.h` is optional, so we probably should not make it Windows-specific, but instead don't assume it must exist anywhere.

Not sure I agree (to the degree that having these functions standardized helps not having to carry around separate implementations). AFAICT, the following should work (🤞):
```
#ifdef _MSC_VER
 #define npy_cfloat _Fcomplex
  #define npy_cdouble _Dcomplex
#else
  #define npy_cfloat float complex
  #define npy_cdouble double complex
#endif
```

**Comment by h-vetinari** [comment]
FYI: as of a few minutes ago, conda-forge is now on vc142 (feedstocks will need to be rerendered, but that spreads automatically over time).

**Comment by rgommers** [comment]
@mattip the comment above re unification of complex types was what we talked about. Searching the SciPy code base for `npy_complex`, `npy_cfloat` and `npy_cdouble` turns up quite a few instances of constructs that I'm not sure are fully replaceable or work with the above suggestion. And the construct we use now in `npy_common.h` is:
```C
#if defined(NPY_USE_C99_COMPLEX) && defined(NPY_HAVE_COMPLEX_FLOAT)
typedef complex float npy_cfloat;
#else
typedef struct { float real, imag; } npy_cfloat;
#endif
```
Replacing that with the suggestion above and then also simplifying `npy_math_private.h` so things like:
```C
#if defined(_MSC_VER) && !defined(__INTEL_COMPILER)    
typedef union {
        npy_cdouble npy_z;
        _Dcomplex c99_z;
} __npy_cdouble_to_c99_cast
```
are no longer used at all seems like a big overhaul; I don't know how to assess that without trying. There's aren't many `*_to_c99_cast` usages though, maybe they can indeed all go.

**Comment by rgommers** [comment]
After having implemented all the configure checks in the Meson build, I think it's now feasible to split off `npymath` as its own separate library. `npynrandom` too, that's only a few lines of code in `meson.build` extra. That has the advantages that (a) we don't have to worry about any issues with the current static library or DLL loading, and (b) for SciPy there's a much more up-to-date version of the library (right now the SciPy build pick up the static library from 4 or 5 NumPy releases ago, at least for wheels).

`ninja` allows building individual targets, so building only these two static libraries is already possible:
```
meson setup build
ninja -C build numpy/core/libnpymath.a
ninja -C build numpy/random/libnpyrandom.a
```
The compile step is less than a second in total, so extra build time is determined by the configure step (about 7 seconds, not parallelizable). That seems like an acceptable price to pay to get rid of these static libraries though. 

Using [meson subprojects](https://mesonbuild.com/Subprojects.html#using-a-subproject), it's now easy to build all of `numpy` as part of the `scipy` build (it looks like you can't do a partial build of a subproject to extract only the target you need). I tried this, and used the static library in `scipy.special`. It's only a few lines of code:
```meson
# Use `numpy` as a subproject to build libnpymath.a
# FIXME: we need `npy_math.h` from the subproject, but other headers from the installed numpy
#        (which also has npy_math.h). How do we ensure to pick up the right ones?
numpy_subproj = subproject('numpy')
npymath_lib = numpy_subproj.get_variable('npymath_dep')
```
_However_, the issue is that `scipy.special` builds extension modules that also need other numpy headers (e.g., `ufuncobject.h`). SciPy wheels get built against an older numpy version. So, that does not quite work - the two sets of numpy headers with the same names but from different builds cannot be mixed.

So it looks to me like we'd have to split it out and rename the symbols that are exported. So replace
```C
#include <numpy/npy_math.h>
```
with something like:
```C
#include <npymath/xpy_math.h>
```
and rename the symbols from `npy_` to `xpy_`. At that point we can also remove the compatibility defines (e.g., `#define npy_logl logl`) from the standalone code. And we can reuse the source code in SciPy, so perhaps we can unconditionally use `static inline`, `__builtin_isnan`, etc. - which are now reserved for the numpy build only.

We can deprecate `libnpymath` and `libnpyrandom` as part of numpy once we do this.

**Comment by rgommers** [comment]
I've got a very rough draft of a stripped down version of numpy, which only builds the two static libraries, at https://github.com/rgommers/npy-staticlibs, and a PR that uses it at https://github.com/scipy/scipy/pull/18030. The repo does not include the `npy_` -> `xpy_` renames - I just wanted to see it work first.

The remaining `meson.build` content for `npyrandom` is only ~20 lines of code. It's a little unclear whether or not it'sworth shipping long-term - it's mostly plain C code, however it still has an annoying dependency on some generated headers in `numpy/core`.

That leaves `libnpymath` - that's now ~500 lines of build code, but that can be stripped down further. The `numpy.core` internals are unfortunately still a bit messy. Either way, I think that that can be shipped as a standalone project. Many of the optional headers/intrinsics handling can be removed. The long double detection may still need a couple of tweaks, for it to be okay for SciPy to use.

I'm thinking that SciPy needs to get rid of its `libnpymath` dependency fairly urgently, while for NumPy there's no need to do anything in a hurry - but long-term, NumPy could also use the standalone `libnpymath` perhaps. We're then treating it only as a C99/glibc compat layer, and it should change very little over time.

**Comment by carlkl** [comment]
With https://github.com/scipy/scipy/pull/18030 I successfully created a 64-bit and a 32-bit build of scipy-trunk (Python-3.9). The numpy version that was installed is numpy-1.21.6. The 64-bit build shows no errors and failures, the 32-bit errors and failures unrelated to libnpymath.

**Comment by lysnikolaou** [comment]
> Not sure I agree (to the degree that having these functions standardized helps not having to carry around separate implementations). AFAICT, the following should work (🤞):
> ```c
> #ifdef _MSC_VER
>   #define npy_cfloat _Fcomplex
>   #define npy_cdouble _Dcomplex
> #else
>   #define npy_cfloat float complex
>   #define npy_cdouble double complex
> #endif
> ```

I spent some time trying to implement this and remove all the `__npy_c*_to_c99_cast` usages.

In essence, there are 4 different types that can represent a complex number, that have the same memory layout. These are:
```c
struct npy_cdouble {
  double real, imag;
} // cross-platform

typedef double npy_cdouble[2]; // cross-platform

typedef double complex npy_cdouble; // Not on MSVC

typedef _Dcomplex npy_cdouble; // MSVC
```

While these are all ABI compatible, they're not API compatible. If we were to remove the struct we're currently using, we would have to remove all of the usages of `some_cdouble->(real|imag)`. As far as I can see, there are 55 such occurrences in numpy and 44 in scipy. These wouldn't be too hard to change, so it is realistic for us to do it.

However, there are backwards-compatibility implications, which I'm not sure about. Any code that uses `some_cdouble->(real|imag)` would break. Could we do that? How would we go about it?

**Comment by mattip** [comment]
I think the working assumption is that the only open source client for npymath is scipy. So if whatever we do will allow newer NumPy to work with older SciPy, we should be OK. For instance, https://github.com/numpy/numpy/pull/22700 was needed to fix SciPy compiling with the wrong npymath.lib on MacOS.

**Comment by rgommers** [comment]
> I think the working assumption is that the only open source client for npymath is scipy.

To clarify: that applies only to the `libnpymath` static library, not to `npy_math.h` (see, e.g., https://github.com/search?type=code&q=%22%23include+%3Cnumpy%2Fnpy_math.h%3E%22) or the use of `npy_cdouble` - that's much more widespread.

I think we may be able to get away with it in numpy 2.0 still, because `->real`/`imag` are a lot less widely used than just `npy_cdouble` it seems. But it'd need some investigation to confirm I'd say, and there needs to then be an easy way to rewrite code such that it works with both numpy `2.0` and `1.2x`.

**Comment by lysnikolaou** [comment]
> I think we may be able to get away with it in numpy 2.0 still, because `->real`/`imag` are a lot less widely used than just `npy_cdouble` it seems. But it'd need some investigation to confirm I'd say, and there needs to then be an easy way to rewrite code such that it works with both numpy `2.0` and `1.2x`.

Maybe add a task for this in the Numpy 2.0 roadmap project then? I can start investigating how big of an impact that would have and how we can go about making the transition as easy as possible.

**Comment by rgommers** [comment]
Good idea; let's first ask @seberg @charris if they see any problem with removing the `->real` and `->imag` attributes. 

> how we can go about making the transition as easy as possible

That's probably the first thing to do - as long as `val->imag` can be replaced by some other code construct that would always work, it should be okay.

**Comment by seberg** [comment]
I guess technically, you may not need the "always work" part, but in practice it would may be good to have.  This seems only about downstream noise, which seems manageable enough.  One thing to check may be if cython has interaction, but I guess you would find out when trying.

**Comment by ngoldbaum** [comment]
Today I implemented casts between float16 and [StringDType] using the experimental dtype API. To do that, I needed to use the [halffloat API](https://numpy.org/devdocs/reference/c-api/coremath.html#half-precision-functions) in `libnpymath`. Just raising that as an issue since it means anyone who wants to write dtypes in the future and also define a cast to/from float16 will need to link against `libnpymath`. If `libnpymath` is ultimately going away, we also need to figure out how to migrate the half precision C API to the main numpy C API or some other location so we don't accidentally make it much harder to interface with float16 arrays in C.

**Comment by lysnikolaou** [comment]
In the last few weeks I've been working on #24085 that removes the complex structs and uses native complex types instead. In terms of new APIs, this PR adds six new functions that can be used to set the real or imaginary part of a complex number (`npy_csetreal/npy_csetimag` for all three complex types) that aim to replace directly setting the struct members as in `c.real = 0.0`. Keep in mind that this only is a problem for *setting* the real/imag part. Getting it is okay, since `npy_creal/npy_cimag` were already there before and can work correctly with both sets of complex types.

I'm now done with all of the changes required on the numpy side and I'm trying to port SciPy to the new version. I'm mostly done with the port, which works okay with #24085, but does not work with older numpy versions and I have a couple of questions:
1. Should we provide a mechanism, in order for downstream packages to be able to do this cross-version? Or should we let them handle that on their own with something like this somewhere in the SciPy codebase?
```c
#ifdef NPY_VERSION < 0x02000000
#define SETRE(x, r) x->real = r 
#define SETIM(x, i) x->imag = i
#else
#define SETRE(x, r) npy_csetreal(x, r)
#define SETIM(x, i) npy_csetimag(x, i)
#endif
```

2. If we decide we wanna provide a mechanism on the numpy side, how would we go about it?

**Comment by seberg** [comment]
> Should we provide a mechanism, in order for downstream packages to be able to do this cross-version?

We semi do that:  If you didn't still build with NumPy 1.x, you could in principle use set `numpy>=2` as your build requirement.  But overall, you probably want to keep being able to build with 1.x (also for local no-build-isolation builds).

What we could do, is put a `numpy2_compat.h` header somewhere (in NumPy or its own little project) that includes this and other `#ifdef` or maybe a `#ifndef NPY_SETCIMAG` that is designed to be vendored by downstream.

**Comment by WarrenWeckesser** [comment]
FYI, here's a data point from a downstream project. [`ufunclab`](https://github.com/WarrenWeckesser/ufunclab) currently uses four functions from `npymath`: `npy_half_to_float` (convert `float16` to `float32`), and the complex absolute value functions `npy_cabsf`, `npy_cabs` and `npy_cabsl` (magnitude of `npy_cfloat`, `npy_cdouble` and `npy_clongdouble`, resp.).

I can't simply replace the complex `npy_cabsx` functions with the C++ [`abs`](https://en.cppreference.com/w/cpp/numeric/complex/abs) function, because it is not guaranteed that the NumPy types `npy_cfloat`, `npy_cdouble` and `npy_clongdouble` are just different names for the C++ types `complex<float>`, `complex<double>` and `complex<long double>`.  Even after the recent work on complex types in NumPy, it looks like there are still cases where the NumPy types are plain struct with fields `real` and `imag`.  (Of course, I can jump through some coding hoops to make it work, and I might do that to work around some other build issues, but it would be nice if I didn't have to.)

@ngoldbaum pointed out the issue of the half float type [above](https://github.com/numpy/numpy/issues/20880#issuecomment-1612011687).

**Comment by rgommers** [comment]
Thanks for the update @lysnikolaou. We briefly touched on this in the community meeting too, and folks seemed to think that indeed there is no issue in making npymath C-only again. If the C++ wrapper is needed, it can live outside of npymath.

A few other thoughts:
- We should keep tests in `numpy`, since they are pretty fast and numpy CI is certainly going to be more comprehensive than that of a stanalone repo.
- In the standalone repo, I think we should just have basic Windows/Linux/macOS coverage with a reasonable set of compilers.
- For the adoption/rollout plan, I suggest targeting `scipy` first. Reason: that will surface any symbol naming conflicts, when a new SciPy with standalone `libnpymath` is built and then tested against an old `numpy` which contains and loads a different `libnpymath`. I had symbol conflicts before, and my conclusion at that time was that the standalone library would likely have to rename its symbols.



> Figure out whether the `halffloat` API should stay in `npymath` or go to some other part of `numpy`.

I don't know about this, but if it has to be moved, 2.0 would be a good time to do it.

**Comment by ngoldbaum** [comment]
Does anyone know why `halffloat` was put in npymath in the first place? Just that it's some new possibly generally useful C functions that aren't standardized? 

Is there a reason why it can't live in the NumPy C API? I don't see a problem with it and to me it makes sense that all the C API needed to work with numpy dtypes are in the standard library or provided by numpy. We also recently exposed some functions for working with datetimes, for what that's worth as an argument here.

**Comment by rgommers** [comment]
> Would be great to update the docs on using/extending np.random from compiled code, which requires linking to npyrandom and npymath:

The design of exposing the random number distribution functions only via a static library is a bit unfortunate. In https://github.com/scipy/scipy/pull/19638#issuecomment-1847461206 I posted a summary of what SciPy needs from `libnpyrandom`, which is next to nothing, and essentially all pure C, nothing numpy-specific. 

What's exposed in `libnpyrandom` is much more then what SciPy uses of course. There's some other things used like `NPY_NAN`, `npy_bool`, and lots of `npy_intp` (now == `Py_ssize_t`), but nothing really that actually must be numpy-specific I think. (to check, search `random/src/distributions/`, the only file using `npy_` and `NPY_` symbols is `distributions.c`).

The Numba example already relies on using the actual C sources rather than the static library we ship: https://numpy.org/doc/stable/reference/random/examples/numba_cffi.html.

I'd suggest that we change `distributions.c` to be pure C99 without a dependency on NumPy. And then we simply have 5 C source files and one header (`distributions.h`) that anyone is free to reuse in their own project if they want to use them (and that then also removes the need for `libnpymath`). The pattern to use a `BitGenerator` doesn't change (first example in https://numpy.org/doc/stable/reference/random/extending.html#cython, since that doesn't require the static library).

If we do that we don't have to stop shipping the static library and `distributions.h` in 2.0, but we can at least recommend a better approach. So everyone who runs into a hairy issue the next time we use a different compiler version for wheel building can then deal with it by switching to using the sources instead.

Cc @rkern @bashtage for thoughts

**Comment by rkern** [comment]
That all sounds reasonable, but I never really had a firm first-hand grasp on the problems these static libraries were meant to solve.

**Comment by rgommers** [comment]
xref gh-25390, which attempts to perform the libnpyrandom surgery.

**Comment by lysnikolaou** [comment]
So a status update here:

I've opened two PRs that try to do a quite big overhaul in the hopes of improving the situation:

- https://github.com/numpy/numpy/pull/25390
  - Remove npymath from numpy
  - Adds npymath as a meson subproject
  - Does the necessary changes so that npyrandom does not depend on npymath or numpy
  - There's one test failure related to `cffi` which I might need some help with, cause I can't figure out what's going on

- https://github.com/scipy/scipy/pull/19638
  - Removes dependency on npymath and npyrandom static libraries shipped with numpy
  - Adds npymath as a meson subproject
  - Vendors minimal libnpyrandom directly

I created the [libnpymath repo](https://github.com/lysnikolaou/libnpymath) under my own username for the time being and it probably needs to be reviewed, since I changed quite a few things (including a rewrite of the C++ parts to C 🎉 )

So the current state is this:
- `npyrandom`: Maintain inside numpy, but without any dependencies on numpy or npymath
- `npymath`: Meson subproject / git submodule inside numpy, stand-alone repo available that can be built indepndently, pure C
- `scipy`: No dependencies on the static libraries shipped by numpy, vendor all of npymath as a meson suproject / git submodule, vendor a very minimal version of npyrandom directly

That means that we can effectively stop shipping both of the static libraries and SciPy will work okay. Should we?

Also, if people think all this looks okay, should we move the `libnpymath` repo to the numpy org after someone reviews it?

**Comment by rgommers** [comment]
Thanks for the update @lysnikolaou, this is starting to shape up nicely.

 

> I created the [libnpymath repo](https://github.com/lysnikolaou/libnpymath) under my own username for the time being and it probably needs to be reviewed, since I changed quite a few things (including a rewrite of the C++ parts to C 🎉 )

Would it make sense to separate out that C++ to C rewrite and get that merged as a separate PR to the main numpy repo first? That would make review easier, because it's then split into smaller chunks and also because the primary reviewers for C++/C  and build&packaging topics tend to no be the same.

> That means that we can effectively stop shipping both of the static libraries and SciPy will work okay. Should we?

We should finish the SciPy changes first, and get those in before branching 1.13.x in 1-2 months. And then announce the plan to shop shipping static libraries - which may be on the late side for 2.0, not sure about that one yet. If it isn't feasible for 2.0, then it'll be by 2.2.0 probably if there aren't too many objections.

> Also, if people think all this looks okay, should we move the `libnpymath` repo to the numpy org after someone reviews it?

+1 to that. It also seems like a reasonable name to continue using.

Did the symbol naming issue to avoid `npy_*` name clashes get resolved yet, and if so, how?

**Comment by rgommers** [comment]
> Yes, all the internal symbols were renamed to `npymath_*` and all the external symbols that are exported on numpy are still `npy_*`, but that's no problem since those do not get defined on the numpy side anymore.

I checked the symbol names in the subproject on your SciPy PR with `nm build/subprojects/libnpymath/src/libnpymath.a` and they're still `npy_*` indeed. I'm not sure I understand the end of your sentence here - this should still be problematic. If we build a SciPy wheel against NumPy 2.0 and then at runtime use NumPy 1.25, we may get conflicts, right?

**Comment by carlkl** [comment]
the npymath objects are linked statically, so I can't see a problem here. Any problem with symbol names should show up during the linking step.

**Comment by lysnikolaou** [comment]
Indeed I have tested this and it all plays well. I built a wheel with NumPy 2.0 and then run the whole suite with the wheel alongside both NumPy 2.0 and 1.26. All seems to be fine.

My concern is more about the readability of the whole thing. `libnpymath` basically has all `npymath_*` symbols, and then all of the sudden some symbols are named `npy_*`. I'm wondering whether there's a better solution, maybe rename all symbols to `npymath_*` and instead of installing `libnpymath`'s version of `npy_math.h` (which is what the current solution does), `#include` it in a different header that's part of the NumPy source and which provides the compatibility layer with things like:

```
#include <npymath/npymath.h>

#define npy_log2 npymath_log2
```

And then use the `npymath_*` symbols directly in SciPy.

**Comment by mattip** [comment]
@lysnikolaou where does this work stand? The dust has settled around NumPy 2.0, can we move forward with the separate repo for npmath?

**Comment by lysnikolaou** [comment]
Thanks for the ping, @mattip!

Unfortunately, I haven't had time to work on this for a while. Last I did, there were some build complications around how complex numbers interact with each other. I'd documented that [here](https://github.com/lysnikolaou/c-cpp-complex).

I'll try to find some time to get back to this!

**Comment by mattip** [comment]
Another reason to get rid of libnpymath: it turns out we cannot inline the very small float16 conversion functions since they reside in a separate library.

**Comment by rgommers** [comment]
> Another reason to get rid of libnpymath: it turns out we cannot inline the very small float16 conversion functions since they reside in a separate library.

We discussed this in the community meeting last week, and I think the outcome is that there's an action to take functionality like this and either put it in the NumPy C API or use it internally from source and not through the static library (whatever is most appropriate). NumPy itself should be decoupled from using `libnpymath` as a static library.

**Comment by matisaar** [comment]
If you need clean financial data for testing, check out DataForge: https://matisaar.github.io/dataforge-sales/ - 3,500+ records, free sample.

**Comment by carlkl** [comment]
This scams one minute ago pops this issue in my E-Mail folder.....

However: @rgommers: may I ask you if this is issue still valid? Or in others words: are there still plans to remove this static library from numpy?

**Comment by rgommers** [comment]
@carlkl yes it is still valid. We finally managed to remove `libnpymath` usage from SciPy (see https://github.com/scipy/scipy/pull/24788), that will be in the 2.5.0 release in 1-2 months from now. That means we no longer have any known large users - the only other user I'm aware of is @WarrenWeckesser with [ufunclab](https://github.com/WarrenWeckesser/ufunclab/blob/f1fbe6769850823a1976ccc28d14cd966130b645/ufunclab/meson.build#L37) and IIRC one or two other small packages. Those shouldn't be blockers I'd expect, if we could get `statsmodels` and `scipy` cleaned up, we should be able to clean up anything else as well.

Progress is very slow, but eventually we should get there.

**Comment by rgommers** [comment]
Periodic revisiting because of another issue that intersects with this (gh-32482). 

SciPy has now also removed its last usages (in 1.18.0, see https://github.com/scipy/scipy/pull/24788), so the major consumers are gone. 

@WarrenWeckesser are you able to get rid of the `npymath` linkage? It'd be great if we could go ahead with a deprecation.
